state/application.go

```diff
@@ -2848,14 +2848,14 @@ func (a *Application) SetConstraints(cons constraints.Value) (err error) {
 	// Note: we don't apply this check in validate constraints as we have no
 	// way to know if it's a model or an application (former we do allow
 	// changes).
-	current, err := a.Constraints()
-	if !errors.IsNotFound(err) && cons.Arch != nil {
+	if current, err := a.Constraints(); !errors.IsNotFound(err) && (cons.Arch != nil && *cons.Arch != "") {
```

Comment:
`err` is shadowing the outer err here so the defer on L2863 will not work as expected if `Contstraints` returns NotFound

Reply:
Good spot, I'll rename the errrrrrr